bank/src/bank_engine.rs

```diff
@@ -838,6 +844,165 @@ impl BankEngine {
+                            };
+                            match tx_state_to_insert.insert(&c) {
+                                Ok(inserted) => inserted,
+                                Err(err) => {
+                                    slog::error!(self.logger, "Could not insert transaction into db: {:?}", err);
+                                    return;
+                                }
+                            }
+                        }
+                    };
+
+                    if db_tx_state.is_settled {
+                        slog::info!(self.logger, "Ignoring already settled transaction: {:?}", tx_state);
+                        return;
+                    }
+
+                    let inbound_uid = match models::on_chain::BitcoinAddress::get_by_address(&c, &tx_state.address) {
+                        Ok(existing_address) => existing_address.uid as u64,
+                        Err(_) => {
+                            slog::warn!(
+                                    self.logger,
+                                    "Could not identify owner of bitcoin address: {}. Ignoring transaction state update: {:?}",
+                                    tx_state.address,
+                                    tx_state);
+                            return;
+                        }
+                    };
+                    let mut inbound_account = {
+                        let user_account = self
+                            .ledger
+                            .user_accounts
+                            .entry(inbound_uid)
+                            .or_insert_with(|| UserAccount::new(inbound_uid));
+
+                        user_account.get_default_account(Currency::BTC)
+                    };
+
+                    let mut external_account = self.ledger.external_account.clone();
+
+                    // Making the transaction and inserting it into the DB.
+                    let amount = Decimal::new(tx_state.value, SATS_DECIMALS);
+                    if self
+                        .make_tx(
+                            &mut external_account,
+                            BANK_UID,
+                            &mut inbound_account,
+                            inbound_uid,
+                            amount,
+                            Decimal::ONE,
+                        )
+                        .is_err()
+                    {
+                        return;
+                    }
+
+                    // Updating cache of external account.
+                    self.ledger.external_account = external_account.clone();
+
+                    // Updating db of internal account.
+                    self.update_account(&inbound_account, inbound_uid);
+
+                    // Updating db of internal account.
+                    self.update_account(&external_account, BANK_UID);
+
+                    if let Err(err) = models::on_chain::OnchainTransaction::set_settled(&c, &tx_state.txid) {
```

Comment:
This is dangerous. If the insertion fails it will be add the funds again on next block. This flag should be set before we make the accounting change. If fails we don't make the accounting change.

Reply:
Well spotted. It has been rectified now.